server/grpc/server.go

```diff
@@ -0,0 +1,151 @@
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+
+	commonv1pb "github.com/dapr/go-sdk/dapr/proto/common/v1"
+	pb "github.com/dapr/go-sdk/dapr/proto/runtime/v1"
+	"github.com/dapr/go-sdk/server/event"
+	"google.golang.org/grpc"
+)
+
+// NewServer creates new Server
+func NewServer(port string) (server *Server, err error) {
+	address := fmt.Sprintf(":%s", port)
+	lis, err := net.Listen("tcp", address)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to TCP listen on: %s", address)
+		return
+	}
+
+	server = &Server{
+		listener:           lis,
+		invokeHandlers:     make(map[string]func(ctx context.Context, contentTypeIn string, dataIn []byte) (contentTypeOut string, dataOut []byte)),
+		topicSubscriptions: make(map[string]func(ctx context.Context, event *event.TopicEvent) error),
+		bindingHandlers:    make(map[string]func(ctx context.Context, in *event.BindingEvent) error),
+	}
+	return
+}
+
+// Server is the gRPC server implementation for Dapr
+type Server struct {
+	listener           net.Listener
+	invokeHandlers     map[string]func(ctx context.Context, contentTypeIn string, dataIn []byte) (contentTypeOut string, dataOut []byte)
+	topicSubscriptions map[string]func(ctx context.Context, event *event.TopicEvent) error
+	bindingHandlers    map[string]func(ctx context.Context, in *event.BindingEvent) error
+}
+
+// Start registers the server and starts it
+func (s *Server) Start() error {
+	gs := grpc.NewServer()
+	pb.RegisterAppCallbackServer(gs, s)
+	return gs.Serve(s.listener)
+}
+
+// START INVOKE
+
+// Invocation
+
+// AddInvocationHandler adds provided handler to the local collection before server start
+func (s *Server) AddInvocationHandler(method string, fn func(ctx context.Context, contentTypeIn string, dataIn []byte) (contentTypeOut string, dataOut []byte)) {
+	s.invokeHandlers[method] = fn
+}
+
+// OnInvoke gets invoked when a remote service has called the app through Dapr
+func (s *Server) OnInvoke(ctx context.Context, in *commonv1pb.InvokeRequest) (*commonv1pb.InvokeResponse, error) {
+	if val, ok := s.invokeHandlers[in.Method]; ok {
+		ct, d := val(ctx, in.ContentType, in.Data.Value)
+		return &commonv1pb.InvokeResponse{
+			ContentType: ct,
+			Data:        &any.Any{Value: d},
+		}, nil
+	}
+	return nil, fmt.Errorf("method not implemented: %s", in.Method)
+}
+
+// START TOPIC SUB
+
+// AddTopicEventHandler adds provided topic to the list of server subscriptions
+func (s *Server) AddTopicEventHandler(topic string, fn func(ctx context.Context, event *event.TopicEvent) error) {
+	s.topicSubscriptions[topic] = fn
```

Comment:
ListTopicSubscriptions can get the topic metadata which allows users to init pubsub component with custom options. Have you considered this metadata support?

Reply:
Ye, I was considering this, probably by adding additional add method (e.g. `AddTopicEventHandlerWithOptions` that would include the metadata. Users implementing simple cases won't burden this way with the additional metadata unless they need it. We can add in future release